misc/amia_paper_2024/stats.py

```diff
@@ -0,0 +1,258 @@
+    pd.options.mode.chained_assignment = None  # default='warn'
+
+    if not INDIR.exists():
+        OUTDIR.mkdir()
+    if not OUTDIR.exists():
+        OUTDIR.mkdir()
+
+
+def api_runs_query(verbose=False):
+    """Get all records from api_runs table"""
+    t0 = datetime.now()
+    with get_db_connection() as con:
+        data: List[RowMapping] = sql_query(
+            con, """
+                SELECT DISTINCT *,
+                date_bin('1 week', timestamp::TIMESTAMP, TIMESTAMP '2023-10-30')::date week,
+                timestamp::date date FROM public.api_runs r""")
+    data: List[Dict] = [dict(d) for d in data]
+    if verbose:
+        print(f'api_runs_query(): Fetched {len(data)} records in n seconds: {(datetime.now() - t0).seconds}')
+    return data
+
+
+def get_dataset_with_mods(func: Callable, path: Union[str, Path], use_cache=False, verbose=True) -> pd.DataFrame:
+    """Get data to be analyzed: JOIN of api_runs and apiruns_grouped
+
+    Mods: Adds client IP column. Filters out null call groups."""
+    if not use_cache:
+        data: List[Dict] = func(verbose)
+        df = pd.DataFrame(data)
+    else:
+        # Fetch data
+        if not os.path.exists(path):
+            # Fetch
+            data: List[Dict] = func(verbose)
+            df = pd.DataFrame(data)
+            # Save
+            t0 = datetime.now()
+            df.to_csv(path, index=False)
+            if verbose:
+                print(f'Wrote {len(df)} records to {path} in n seconds: {(datetime.now() - t0).seconds}')
+        else:
+            # Read
+            t0 = datetime.now()
+            df = pd.read_csv(path)
+            if verbose:
+                print(f'Read {len(df)} records from {path} in n seconds: {(datetime.now() - t0).seconds}')
+    # Mods
+    df['client_ip'] = df['client'].apply(lambda x: x.split(':')[0])
+    # Filtering
+    df = preprocess_null_call_groups(df, verbose)
+    # Formatting
+    df = df.fillna('')  # fixes problem w/ .str.contains() ops
+    return df
+
+
+def filter_dev_data(df: pd.DataFrame, verbose=True) -> pd.DataFrame:
+    """Preprocess data: remove developers"""
+    # Devs accessing dev/prod
+    developer_ips = {'Siggie': '216.164.48.98', 'Joe': '174.99.54.40'}
+    if verbose:
+        print('Filtered out n records based on IP addresses of each developer accesing dev/prod:')
+        for dev, ip in developer_ips.items():
+            df_i = df[df['client_ip'] == ip]
+            print(f'  - {dev}: {len(df_i)}')
+    df2 = df[~df['client_ip'].isin(list(developer_ips.values()))]
+    # Devs running locally
+    count_before = len(df2)
+    df2 = df2[df2['host'].isin(['dev', 'prod'])]
+    if verbose:
+        print('Filtered out n records based devs running locally: ',  count_before - len(df2))
+    # Test cases being run by GitHub actions: IPs unknown
+    test_cases = [
+        '[1000002363, 1000002657, 1000007602, 1000013397, 1000010688, 1000015307, 1000031299]',  # many small
```

Comment:
## Filter: automated test logs
@Sigfried I am filtering these out now. This was about 33% of the data.

I did a face check and confirmed as well by looking at the timestamps. All of these removals I saw were around 3AM our time, which is what I'd expect from the cronjob:
> `- cron: '0 8 * * *'  # every day 8am GMT (3-4am EST/EDT)`